core/src/main/scala/org/http4s/Headers.scala

```diff
@@ -5,28 +5,27 @@ import cats.implicits._
 import org.http4s.headers.`Set-Cookie`
 import org.http4s.syntax.string._
 import org.http4s.util.CaseInsensitiveString
-import scala.collection.generic.CanBuildFrom
 import scala.collection.mutable.ListBuffer
-import scala.collection.{GenTraversableOnce, immutable, mutable}
 
 /** A collection of HTTP Headers */
-final class Headers private (headers: List[Header])
-    extends immutable.Iterable[Header]
-    with collection.IterableLike[Header, Headers] {
+final class Headers private (private val headers: List[Header]) extends AnyVal {
 
-  override def toList: List[Header] = headers
+  def toList: List[Header] = headers
 
-  override def isEmpty: Boolean = headers.isEmpty
+  def isEmpty: Boolean = headers.isEmpty
 
-  override protected def newBuilder: mutable.Builder[Header, Headers] = Headers.newBuilder
+  def nonEmpty: Boolean = headers.nonEmpty
 
-  override def drop(n: Int): Headers = if (n == 0) this else new Headers(headers.drop(n))
+  def drop(n: Int): Headers = if (n == 0) this else new Headers(headers.drop(n))
 
-  override def head: Header = headers.head
+  def iterator: Iterator[Header] = headers.iterator
 
-  override def foreach[B](f: Header => B): Unit = headers.foreach(f)
+  def tail: Headers = Headers(headers match {
+    case _ :: xs => xs
+    case n@Nil => n 
+  })
 
-  def iterator: Iterator[Header] = headers.iterator
+  def head: Header = headers.headOption.fold[Header](throw new NoSuchElementException)(identity)
```

Comment:
I think the only reason to have partial methods was to be a proper Scala collection.  If we're abandoning that goal, I would call this `headOption`.

Now, for backward compatibility, we could go nuts with deprecation on all the collection methods we've left behind, or we could pick a few major ones like this.

Reply:
This was being used for a fully tailrecursive method downstream, however we can instead convert to list in that location.